Application/EdFi.Ods.Tests/EdFi.Ods.Api/Controllers/TokenControllerTests.cs

```diff
@@ -1,1626 +1,1652 @@
-    //             public void Should_not_include_any_cache_headers()
-    //             {
-    //                 Assert.That(_actualResponseMessage.Headers.CacheControl, Is.Null);
-    //                 Assert.That(_actualResponseMessage.Headers.Pragma, Is.Empty);
-    //             }
-    //
-    //             public override void RunOnceAfterAll()
-    //             {
-    //                 _clientAppRepo.VerifyAllExpectations();
-    //             }
-    //         }
-    //     }
-    // }
+    public class When_calling_the_token_controller
+    {
+
+        private static TokenController CreateTokenController(
+            ClientCredentialsTokenRequestProvider tokenRequestProvider)
+        {
+
+            var controller = new TokenController(tokenRequestProvider);
+            var request = A.Fake<HttpRequest>();
+            request.Method = "Post";
+            request.Scheme = "http";
+
+            A.CallTo(() => request.Host).Returns(new HostString("localhost", 80));
+
+            request.PathBase = "/";
+            request.RouteValues = new RouteValueDictionary { {
+                            "controller", "authorize"
+                        } };
+
+            var httpContext = A.Fake<HttpContext>();
+            A.CallTo(() => httpContext.Request).Returns(request);
+
+            var controllerContext = new ControllerContext()
+            {
+                HttpContext = httpContext,
+            };
+            var routeData = A.Fake<RouteData>();
+            RouteValueDictionary dictionary = new RouteValueDictionary();
+            dictionary.Add("controller", "authorize");
+
+            controllerContext.RouteData = new RouteData(dictionary);
+            controller.ControllerContext = controllerContext;
+
+            return controller;
+        }
+
+        public class Using_client_credentials_grant
+        {
+            [TestFixture]
+            public class With_valid_key_and_secret : TestFixtureBase
+            {
+                private IClientAppRepo _clientAppRepo;
+                private IApiClientAuthenticator _apiClientAuthenticator;
+                private TokenController _controller;
+
+                private ApiClient _suppliedClient;
+                private Guid _suppliedAccessToken;
+
+                private TimeSpan _suppliedTTL;
+                private IActionResult _actionResult;
+
+                protected override void Arrange()
+                {
+                    _suppliedClient = new ApiClient
+                    {
+                        ApiClientId = 1
+
+                    };
+
+                    _suppliedAccessToken = Guid.NewGuid();
+                    _suppliedTTL = TimeSpan.FromMinutes(30);
+
+                    _clientAppRepo = Stub<IClientAppRepo>();
+
+                    //// Simulate a successful lookup of the client id/secret
+                    A.CallTo(() => _clientAppRepo.GetClientAsync(A<string>._)).Returns(_suppliedClient);
```

Comment:
I think we have an issue here, we are using async methods, and you should be returning a Task.FromResult(). Secondly the test base framework is not designed to be async, and we should consider rewritting these tests using the AAA syntax and using  async.